File: server/src/main/java/com/krillsson/sysapi/core/monitoring/monitors/DriveMonitor.java
package com.krillsson.sysapi.core.monitoring.monitors;

import com.krillsson.sysapi.core.domain.system.SystemLoad;
import com.krillsson.sysapi.core.monitoring.Monitor;

import java.time.Duration;

import static com.krillsson.sysapi.core.metrics.Empty.DRIVE_LOAD;


public class DriveMonitor extends Monitor {
    public DriveMonitor(String id, Duration inertia, double threshold) {
        super(id, inertia, threshold);
    }

    @Override
    protected double value(SystemLoad systemLoad) {
        return (double) systemLoad.getDriveLoads()
                .stream()
                .filter(i -> i.getName().equalsIgnoreCase(id()))
                .findFirst().orElse(DRIVE_LOAD)
                .getMetrics()
                .getUsableSpace();
    }

    @Override
    protected boolean isOutsideThreshold(double value) {
        return value < threshold();
    }

    @Override
    protected MonitorType type() {
        return MonitorType.DRIVE;
    }

    @Override
    public String toString() {
        return "DriveMonitor{" +
                "threshold=" + threshold() +
                '}';
    }
}

File: server/src/main/java/com/krillsson/sysapi/core/monitoring/monitors/DriveMonitor.kt
package com.krillsson.sysapi.core.monitoring.monitors

import com.krillsson.sysapi.core.domain.drives.DriveLoad
import com.krillsson.sysapi.core.domain.system.SystemLoad
import com.krillsson.sysapi.core.metrics.Empty
import com.krillsson.sysapi.core.monitoring.Monitor
import com.krillsson.sysapi.core.monitoring.MonitorType
import java.util.*

class DriveMonitor(override val id: UUID, override val config: Config) : Monitor() {
    override val type: MonitorType = MonitorType.DRIVE_SPACE

    override fun selectValue(load: SystemLoad): Double = load.driveLoads
            .stream()
            .filter { i: DriveLoad -> i.serial.equals(config.id, ignoreCase = true) || i.name.equals(config.id, ignoreCase = true) }
            .findFirst()
            .orElse(Empty.DRIVE_LOAD)
            .values
            .usableSpace.toDouble()

    override fun isPastThreshold(value: Double): Boolean {
        return value < config.threshold
    }
}